### PR TITLE
[SPARK-31017][TEST][CORE] Test for shuffle requests packaging with different size and numBlocks limit

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

Added 2 tests for `ShuffleBlockFetcherIteratorSuite`.

### Why are the changes needed?

When packaging shuffle fetch requests in `ShuffleBlockFetcherIterator`, there are two limitations: `maxBytesInFlight` and `maxBlocksInFlightPerAddress`. However, we don’t have test cases to test them both, e.g. the size limitation is hit before the numBlocks limitation.

We should add test cases in `ShuffleBlockFetcherIteratorSuite` to test:

1. the size limitation is hit before the numBlocks limitation
2. the numBlocks limitation is hit before the size limitation

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added new tests.

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan , please take a look, thanks!

**Comment by Ngone51** [comment]
Jenkins, retest this please.

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 1]
```diff
@@ -254,6 +254,78 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
...
 
+  test("SPARK-31017: Hit maxBytesInFlight limitation before maxBlocksInFlightPerAddress") {
+    val blockManager = mock(classOf[BlockManager])
+    val localBmId = BlockManagerId("test-client", "test-local-host", 1)
+    doReturn(localBmId).when(blockManager).blockManagerId
+
+    val remoteBmId1 = BlockManagerId("test-remote-client-1", "test-remote-host1", 1)
+    val remoteBmId2 = BlockManagerId("test-remote-client-2", "test-remote-host2", 2)
+    val remoteBmId3 = BlockManagerId("test-remote-client-3", "test-remote-host2", 3)
+    // set maxBlocksInFlightPerAddress to Int.MaxValue,
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala]: seems we don't need this line, as we have it in https://github.com/apache/spark/pull/27767/files#diff-d9abbfedd9c2b0aed76c94c91ec0cec2R283

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 2]
```diff
@@ -254,6 +254,78 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
...
+    verify(transfer, times(2)).fetchBlocks(any(), any(), any(), any(), any(), any())
+  }
+
+  test("SPARK-31017: Hit maxBlocksInFlightPerAddress limitation before maxBytesInFlight") {
+    val blockManager = mock(classOf[BlockManager])
+    val localBmId = BlockManagerId("test-client", "test-local-host", 1)
+    doReturn(localBmId).when(blockManager).blockManagerId
+
+    val remoteBmId = BlockManagerId("test-remote-client-1", "test-remote-host", 2)
+    // set maxBlocksInFlightPerAddress to 2, so there will be 2 FetchRequests after initialize.
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala]: can we move this comment to https://github.com/apache/spark/pull/27767/files#diff-d9abbfedd9c2b0aed76c94c91ec0cec2R318 ?

**Comment by cloud-fan** [comment]
The tests look great! Can we refactor the test for continuous fetch "fetch continuous blocks in batch respects maxSize and maxBlocks" so that it can test hitting these 2 limitations separately?

File: core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala [hunk 3]
```diff
@@ -341,31 +433,85 @@ class ShuffleBlockFetcherIteratorSuite extends SparkFunSuite with PrivateMethodT
...
+      val mockBuf = mergedRemoteBlocks(blockId)
+      verifyBufferRelease(mockBuf, inputStream)
+      numResults += 1
+    }
+    // The last request will be sent after next() is called.
+    verify(transfer, times(6)).fetchBlocks(any(), any(), any(), any(), any(), any())
+    assert(numResults == 6)
+  }
+
+  // TODO(wuyi) test this after the fix
```
Comment (Ngone51) [inline core/src/test/scala/org/apache/spark/storage/ShuffleBlockFetcherIteratorSuite.scala]: should be fixed by #27786.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
This just adds 2 tests and that test suite has passed in the last jenkins run.

Thanks, merging to master/3.0!

**Comment by Ngone51** [comment]
thanks!

**Pull request closed without merging**